lib/storage/index.js

```diff
@@ -0,0 +1,215 @@
+import PlatformStorage from './platforms';
+import NoopProvider from './providers/NoopProvider';
+import InstanceSync from './InstanceSync';
+import * as Logger from '../Logger';
+
+let provider = PlatformStorage;
+let finishInitalization;
+const initPromise = new Promise((resolve) => {
+    finishInitalization = resolve;
+});
+let shouldKeepInstancesSync = false;
+
+/**
+ * Runs an operation after the storage layer has been initialized
+ * @param {Function} callback - a callback that should be run after initialization
+ * @param {Array<any>} params - the parameters to pass to the operation
+ * @return {Promise<void>}
+ */
+const runAfterInit = (callback, ...params) =>
+    new Promise((resolve, reject) =>
+        initPromise.then(() =>
+            callback(...params)
+                .then(resolve)
+                .catch(reject),
+        ),
+    );
+
+const Storage = {
+    getStorageProvider() {
+        return provider;
+    },
+
+    /**
+     * Initializes all providers in the list of storage providers
+     * and enables fallback providers if necessary
+     * @return {Boolean} whether a fallback provider is being used
+     */
+    init() {
+        // If a storage provider fails to initialize, we remove it from the list of providers
+        // We don't want to remove the "memory only" provider though, therefore, we start at index 1
+        let isUsingFallback = false;
+
+        try {
+            provider.init();
+        } catch (error) {
+            Logger.logAlert(`${provider.name} could not be initialized. Falling back to the memory-only provider instead.`);
+            console.error(error);
+
+            provider = NoopProvider;
+            isUsingFallback = true;
+        }
+
+        finishInitalization();
+
+        return isUsingFallback;
+    },
+
+    /**
+     * Get the value of a given key or return `null` if it's not available in memory
+     * @param {String} key
+     * @return {Promise<*>}
+     */
+    getItem(key) {
+        return runAfterInit(() => provider.getItem(key));
```

Comment:
**confusion:** Why do we need `runAfterInit()`?

Reply:
Since i moved the initialization of the provider to `provider.init()`, we need this, to prevent any operations from running before the storage provider actually got intialized/created